daemon/daemon.go

```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	return nil
 }
 
+func writeDistributionProgress(cancelFunc func(), outStream io.Writer, progressChan <-chan xfer.Progress) {
+	sf := streamformatter.NewJSONStreamFormatter()
+	for prog := range progressChan {
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
```

Comment:
> The function needs to continue to consume messages from progressChan until it's closed, otherwise the writer could block forever.

To clarify this: cancellation is asynchronous, so there can be a race between cancelling a push or pull, and getting further progress messages.